File: bot/util/cache.py
import orjson
from redis.asyncio import Redis

from bot.youtube import VideoInfo


class VideoInfoCache:
    def __init__(self, redis: Redis):
        self.redis = redis

    async def get(self, key: str) -> VideoInfo | None:
        data = await self.redis.get(key)
        if data is None:
            return None
        print("Hit cache for ", key)
        return VideoInfo(**orjson.loads(data))

    async def set(self, info: VideoInfo):
        return await self.redis.set(
            info.url,
            orjson.dumps(info),
        )

File: src/cache.rs
use redis::AsyncCommands;
use crate::youtube::VideoInfo;

pub struct VideoInfoCache<'a> {
    client: &'a redis::Client,
}

impl<'a> VideoInfoCache<'a> {
    pub fn new(client: &'a redis::Client) -> Self {
        Self { client }
    }

    pub async fn get(&self, url: &str) -> Option<VideoInfo> {
        let mut conn = self.client.get_multiplexed_async_connection().await.ok()?;
        let data: String = conn.get(url).await.ok()?;
        println!("Hit cache for {}", url);
        serde_json::from_str(&data).ok()
    }

    pub async fn set(&self, info: &VideoInfo) -> anyhow::Result<()> {
        let mut conn = self.client.get_multiplexed_async_connection().await?;
        let data = serde_json::to_string(info)?;
        let _: () = conn.set(info.url.clone(), data).await?;
        Ok(())
    }
}
